WordPress/gutenberg
# StartPageOptions: load and parse patterns only after establishing the need for them

This PR fixes an issue where block patterns are parsed during the editor initial load, synchronously, and therefore are slowing down the load. I discovered the issue when working on async block loading in #51778.

The problem is with the `<StartPageOptions>` component (added in #40034) which displays a modal to select a pattern when creating a brand new page from scratch. This component looks at the list of patterns, which triggers pattern parsing. But most of the time the component will not be displayed at all, because there are preconditions that are rarely satisfied: the page content is empty and unsaved. And yet patterns get synchronously parsed during the initial load.

This PR fixes the issue by refactoring `StartPageOptions` into two nested components. The outer one checks the preconditions (freshly created page) and renders the inner component, the one that parses patterns, only when the precondition is satisfied.

The second commit removes the `usePreParsePatterns` hook (added in #29444) completely. Even though this hook parses patterns only during idle time, I believe we don't need it. It's OK to wait until patterns are actually needed, e.g., when opening the block inserter.

**How to test:**
Verify that before this PR patterns were parsed on initial load, while after this PR they are parsed only when opening the block inserter. Can be done by adding a logpoint to the `__experimentalGetParsedPattern` selector.

Then verify that `StartPageOptions` is correctly shown. It's shown after creating a new page, and when you have a pattern with `blockTypes = [ 'core/post-content' ]`. You can temporarily change the condition from `core/post-content` to `core/query` to see matching patterns in themes like Twenty Twenty Three.
